File: windows/DMShot/Video/GifPlan.cs
namespace DMShot.Video;

/// <summary>
/// The two selectable GIF output levels (spec 2026-07-05, mac parity: GIFQuality).
/// Standard is the unchanged default pipeline; Small trades fidelity for roughly a
/// quarter of the size (half the frames × ~2/3 the pixels).
/// </summary>
public enum GifQuality { Standard, Small }

public static class GifQualityExtensions
{
    public static double Fps(this GifQuality q) => q == GifQuality.Small ? 5.0 : GifPlan.DefaultFps;
    public static int MaxWidth(this GifQuality q) => q == GifQuality.Small ? 800 : GifPlan.DefaultMaxWidth;
}

/// <summary>
/// Pure planning math for GIF encoding (no I/O). Shared contract for both platforms;
/// a direct port of the macOS GIFPlan.swift. bytesPerPixelPerFrame is tuned to 0.25
/// because frame-dedup collapses the (usually large) static regions of screen recordings.
/// </summary>
public static class GifPlan
{
    public const double DefaultFps = 10.0;
    public const int DefaultMaxWidth = 1000;
    public const double BytesPerPixelPerFrame = 0.25;

    public static double[] FrameTimes(double durationSec, double fps = DefaultFps)
    {
        int count = Math.Max(1, (int)Math.Round(durationSec * fps));
        var t = new double[count];
        for (int i = 0; i < count; i++) t[i] = i / fps;
        return t;
    }

    public static (int W, int H) ScaledSize(int width, int height, int maxWidth = DefaultMaxWidth)
    {
        if (width <= maxWidth || width <= 0) return (width, height);
        double scale = (double)maxWidth / width;
        return (maxWidth, Math.Max(1, (int)Math.Round(height * scale)));
    }

    public static long EstimatedBytes(int frameCount, int width, int height)
        => (long)((double)frameCount * width * height * BytesPerPixelPerFrame);
}

File: mac/Sources/DMShot/GIFPlan.swift
import Foundation

/// Pure planning math for GIF encoding (no I/O). Shared contract for both platforms.
enum GIFPlan {
    static let defaultFPS: Double = 10
    static let defaultMaxWidth: Int = 1000
    /// Rough average compressed bytes per output pixel per frame. Tuned down from a
    /// naive 0.5 because whole-static frames are collapsed upstream by the dedup pass
    /// and LZW compresses the mostly-flat content of screen recordings well. (The mac
    /// encoder writes full frames — no inter-frame delta optimization; Windows'
    /// ImageSharp additionally delta-crops frames, see PARITY.md step 6.)
    static let bytesPerPixelPerFrame: Double = 0.25

    /// Sample times (seconds, relative to range start) for `duration` at `fps`.
    static func frameTimes(duration: Double, fps: Double = defaultFPS) -> [Double] {
        let count = max(1, Int((duration * fps).rounded()))
        return (0..<count).map { Double($0) / fps }
    }

    /// Scale `width`x`height` down so width ≤ `maxWidth`, preserving aspect ratio.
    static func scaledSize(width: Int, height: Int,
                           maxWidth: Int = defaultMaxWidth) -> (width: Int, height: Int) {
        guard width > maxWidth, width > 0 else { return (width, height) }
        let scale = Double(maxWidth) / Double(width)
        return (maxWidth, max(1, Int((Double(height) * scale).rounded())))
    }

    /// Rough size estimate (bytes) for the preview readout.
    static func estimatedBytes(frameCount: Int, width: Int, height: Int) -> Int {
        Int(Double(frameCount * width * height) * bytesPerPixelPerFrame)
    }
}
